src/audio/kpb.c

```diff
@@ -2378,22 +2398,202 @@ static int kpb_set_micselect(struct comp_dev *dev, const void *data,
+		return -EINVAL;
+
+	int ret = 0;
+	struct comp_dev *dev = NULL;
+	struct kpb_fmt_dev_list *fmt_device_list =
+		&((struct comp_data *)comp_get_drvdata(kpb_dev))->fmt_device_list;
+
+	if (modules_to_prepare->number_of_modules != 0) {
+		fmt_device_list->kpb_list_item[outpin_idx] = kpb_dev;
+		ret = devicelist_push(&fmt_device_list->device_list[outpin_idx],
+				      &fmt_device_list->kpb_list_item[outpin_idx]);
+		if (ret < 0)
+			return ret;
+	}
+	for (size_t mod_idx = 0; mod_idx < modules_to_prepare->number_of_modules; ++mod_idx) {
+#ifdef CONFIG_IPC_MAJOR_4
+		uint32_t comp_id = IPC4_COMP_ID(modules_to_prepare->dev_ids[mod_idx].module_id,
+						modules_to_prepare->dev_ids[mod_idx].instance_id);
+
+		dev = ipc4_get_comp_dev(comp_id);
+#endif
+
+		if (!dev)
+			return -EINVAL;
+
+		struct comp_dev **new_list_item_ptr;
+
+		ret = alloc_fmt_module_list_item(fmt_device_list, dev, &new_list_item_ptr);
+		if (ret < 0)
+			return ret;
+		*new_list_item_ptr = dev;
+		ret = devicelist_push(&fmt_device_list->device_list[outpin_idx],
+				      new_list_item_ptr);
+		if (ret < 0)
+			return ret;
+	}
+	return ret;
+}
+
+static void clear_fmt_modules_list(struct kpb_fmt_dev_list *fmt_device_list,
+				   uint32_t outpin_idx)
+{
+	/* Note: this should be validated in layer above. */
+	assert(outpin_idx < KPB_MAX_SINK_CNT);
+
+	devicelist_reset(&fmt_device_list->device_list[outpin_idx], true);
+}
+
+static int unregister_modules_list(struct fast_mode_task *fmt,
+				   struct device_list *list_to_remove, size_t list_idx)
+{
+	if (list_to_remove == fmt->device_list[list_idx]) {
+		fmt->device_list[list_idx] = NULL;
+		return 0;
+	}
+	if (!fmt->device_list[list_idx]) {
+		/* Nothing to do here */
+		return 0;
+	}
+	return -EINVAL;
+}
+
+/* Comment from Old FW, may be outdated:
+ * Important: function below should be called only from within critical section
+ * (Goto KPB for more details)
+ */
+static int register_modules_list(struct fast_mode_task *fmt,
+				 struct device_list *new_list, size_t list_idx)
+{
+	if (list_idx >= ARRAY_SIZE(fmt->device_list))
+		return -EINVAL;
+
+	/* Check if slot is free */
+	if (!fmt->device_list[list_idx]) {
+		fmt->device_list[list_idx] = new_list;
+		return 0;
+	}
+	if (new_list == fmt->device_list[list_idx]) {
+		/* Already registered. */
+		return 0;
+	}
+	/* was ADSP_ALREADY_IN_USE */
+	return -EINVAL;
+}
+
+static int configure_fast_mode_task(struct comp_dev *kpb_dev, const struct kpb_task_params *cfg,
+				    size_t pin)
+{
+	if (pin >= KPB_MAX_SINK_CNT || pin == REALTIME_PIN_ID || !cfg)
+		return -EINVAL;
+
+	struct comp_data *priv_data = (struct comp_data *)comp_get_drvdata(kpb_dev);
+	int ret = unregister_modules_list(&priv_data->fmt,
+					  &priv_data->fmt_device_list.device_list[pin],
+					  pin);
+	if (ret)
+		return -EINVAL;
+
+	clear_fmt_modules_list(&priv_data->fmt_device_list, pin);
+
+	/* When modules count IS 0 we only need to remove modules from Fast Mode. */
+	if (cfg->number_of_modules > 0) {
```

Comment:
the function itself also has checks for `cfg->number_of_modules > 0`  and does nothing if that isn't satisfied. I'm remove this and other already known to hold checks from the function - see comment there

Reply:
Left this one, removed other ones.